extensions/RabbitMQ/RabbitMQPipeline.cs

```diff
@@ -112,23 +126,50 @@ public void OnDequeue(Func<string, Task<bool>> processMessageAction)
-                string message = Encoding.UTF8.GetString(body);
+                if (args.BasicProperties.Headers.TryGetValue(DeliveryCountHeader, out var deliveryCountObj))
+                {
+                    // The number of dequeue attempts is equal to the number of times the message has been delivered +1.
+                    dequeueCount = Convert.ToInt32(deliveryCountObj, CultureInfo.InvariantCulture) + 1;
+                }
 
-                bool success = await processMessageAction.Invoke(message).ConfigureAwait(false);
-                if (success)
+                if (dequeueCount <= this._config.MaxRetriesBeforePoisonQueue)
                 {
-                    this._log.LogTrace("Message '{0}' successfully processed, deleting message", args.BasicProperties.MessageId);
-                    this._channel.BasicAck(args.DeliveryTag, multiple: false);
+                    this._log.LogDebug("Message '{0}' received, expires after {1}ms", args.BasicProperties.MessageId, args.BasicProperties.Expiration);
+
+                    byte[] body = args.Body.ToArray();
+                    string message = Encoding.UTF8.GetString(body);
+
+                    bool success = await processMessageAction.Invoke(message).ConfigureAwait(false);
+                    if (success)
+                    {
+                        this._log.LogTrace("Message '{0}' successfully processed, deleting message", args.BasicProperties.MessageId);
+                        this._channel.BasicAck(args.DeliveryTag, multiple: false);
+                    }
+                    else
+                    {
+                        var backoffDelay = TimeSpan.FromSeconds(1 * dequeueCount);
+
+                        this._log.LogWarning("Message '{0}' failed to process, putting message back in the queue with a delay of {1} msecs",
+                            args.BasicProperties.MessageId, backoffDelay.TotalMilliseconds);
+
+                        await Task.Delay(backoffDelay).ConfigureAwait(false);
+
+                        this._channel.BasicNack(args.DeliveryTag, multiple: false, requeue: false);
+                        this.PublishMessage(this._queueName, args.Body, args.BasicProperties.MessageId, this._messageTTLMsecs, dequeueCount);
+                    }
                 }
                 else
                 {
-                    this._log.LogWarning("Message '{0}' failed to process, putting message back in the queue", args.BasicProperties.MessageId);
-                    this._channel.BasicNack(args.DeliveryTag, multiple: false, requeue: true);
+                    this._log.LogWarning("Message '{0}' has reached the maximum number of retries, moving to poison queue", args.BasicProperties.MessageId);
+
+                    this._channel.BasicNack(args.DeliveryTag, multiple: false, requeue: false);
+                    this.PublishMessage(this._poisonQueueName, args.Body, args.BasicProperties.MessageId, expirationMsecs: null, dequeueCount);
```

Comment:
I thought we were going to use quorum queues, leaving it to RabbitMQ to manage the queue transaction. 

Doing it with code is not 100% reliable:

* if `PublishMessage` runs **after** `BasicNack`
  * if `BasicNack` succeeds and `PublishMessage` fails, the message is lost, and there won't be any retry
* if `PublishMessage` runs **before** `BasicNack`
  * if `PublishMessage` succeeds and `BasicNack` fails, the message is duplicated. If the failure continues duplicate messages will build up, potentially bringing down the service

Reply:
reference for dead letter queues: https://www.rabbitmq.com/docs/quorum-queues#poison-message-handling

Reply:
Actually, I already proposed to use Quorum Queues, but you discarded the idea in https://github.com/microsoft/kernel-memory/issues/408#issuecomment-2099251014, mainly because they don't support a delayed delivery.

Giving that, I can modify this PR to use Quorum Queues instead, without the delayed retry logic.

Reply:
the other comment was about the retry strategy while the PR here is about poison queues. I think there's some misunderstanding, also because of rabbitmq complexity, about what missing features 
* increasing backoff while retrying
* discard messages after N retries
* store discarded messages in poison queues

Reply:
Yes, I agree that RabbitMQ introduces some complexities and I think that we need to better define the requirements, because in this moment I can't get how to use Quorum Queues together with increasing backoff while retrying and the your other points.